File: lib/dart/section_3dot2/diamond_problem_resolution.dart
/// This file includes code examples regarding the resolution of the diamond problem
/// via mixin linearization seen in section 3.2 of the paper.

// The first mixin
mixin Walker {
  String move() => "Walking";
}

// The second mixin
mixin Flyer {
  String move() => "Flying";
}

// A class applying both mixins.
//
// Dart resolves the "Diamond Problem" via linearization.
// Since `Flyer` is applied last (it is the right-most mixin),
// its implementation of `move` overrides `Walker`'s implementation.
class Duck with Walker, Flyer {}

// Class with order swapped for demonstration that order does indeed matter.
class Penguin with Flyer, Walker {}

/// This function exists to showcase basic usage, and also exists to ensure the linter is
/// satisfied regarding dead code. It also shows insightful output when running tests with
/// output visible.
void basic_usage() {
  print("\nRunning mixin_linearization.dart usage example!");
  var duck = Duck();
	var penguin = Penguin();

  // This will print "Flying" because Flyer wins the linearization
  print("Duck says: ${duck.move()}");
  // This will print "Walking" because Flyer wins the linearization
  print("Penguin says: ${penguin.move()}");
  print("");
}

File: lib/rust/section_3dot2/diamond_problem_resolution.rs
/// This file includes code examples regarding the resolution of the diamond
/// problem seen in section 3.2 of the paper.

// A basic walker trait
trait Walker {
    // we use `move_` because `move` is a keyword in rust
    fn move_(&self) -> &'static str {
        "Walking"
    }
}

// A basic flyer trait
trait Flyer {
    fn move_(&self) -> &'static str {
        "Flying"
    }
}

struct Duck;
impl Walker for Duck {}
impl Flyer for Duck {}

/// This function exists to showcase basic usage, and also exists to ensure the linter is
/// satisfied regarding dead code. It also shows insightful output when running tests with
/// output visible.
pub fn basic_usage() {
    println!("\nRunning diamond_problem.rs usage example!");
    let duck = Duck;

    // No ambiguity; must be explicit
    println!("{}", Walker::move_(&duck));
    println!("{}", Flyer::move_(&duck));
    println!();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_trait_method_dispatch() {
        let duck = Duck;
        assert_eq!(Walker::move_(&duck), "Walking");
        assert_eq!(Flyer::move_(&duck), "Flying");
    }

    #[test]
    fn test_basic_usage_runs() {
        // This test simply runs the basic_usage function to ensure it executes without
        // panicking, and enables viewing output when running with output using
        // the method mentioned in the readme.
        basic_usage();
    }
}
